### PR TITLE
Some optimizations

### PR DESCRIPTION
**Pull request by cupertank**

### CODE REVIEW

File: inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt [hunk 1]
```diff
@@ -44,26 +44,8 @@ class BiasGeluVer1(name: String, attributes: Map<String, Attribute<Any>> = empty
...
-                val absValue = abs(valueForErf)
-                val t = 1f / (1f + ERF_P_VALUE_FLOAT * absValue)
-                val sum = t * (ERF_COEF_1_FLOAT + t * (ERF_COEF_2_FLOAT + t * (ERF_COEF_3_FLOAT + t * (ERF_COEF_4_FLOAT + t * ERF_COEF_5_FLOAT))))
-                val erfValue = sign * (1f - sum * exp(-absValue * absValue))
-                0.5f * value * (1.0f + erfValue)
-            }
-        }
+        private val SQRT_2 = sqrt(2f)
 
         private fun computeDouble(output: DoubleNDArray, input: DoubleNDArray, bias: DoubleNDArray, startOffset: Int = 0, count: Int = input.linearSize - startOffset) {
```
Comment (AnastasiaTuchina) [inline inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt:50]: Add the optimizations for double, please...

File: inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt [hunk 2]
```diff
@@ -109,11 +91,89 @@ class BiasGeluVer1(name: String, attributes: Map<String, Attribute<Any>> = empty
...
+            val output = MutableFloatNDArray(input.strides)
+
+            val blocks = input.array.blocks
+            val biasBlocks = bias.array.blocks
+            val outputBlocks = output.array.blocks
+
+            val blockSize = input.array.blockSize
+            val batchSize = run {
+                var batchSize = 1
+                //1720
```
Comment (AnastasiaTuchina) [inline inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt:105]: 1720 what?

File: inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt [hunk 3]
```diff
@@ -109,11 +91,89 @@ class BiasGeluVer1(name: String, attributes: Map<String, Attribute<Any>> = empty
...
+
+        suspend fun computeGelu(input: FloatNDArray, bias: FloatNDArray): MutableFloatNDArray {
+            val output = MutableFloatNDArray(input.strides)
+
+            val blocks = input.array.blocks
+            val biasBlocks = bias.array.blocks
+            val outputBlocks = output.array.blocks
+
+            val blockSize = input.array.blockSize
+            val batchSize = run {
```
Comment (AnastasiaTuchina) [inline inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/math/BiasGelu.kt:103]: Actually, all of these block size functions follow the same pattern. Maybe you can put this as a function somewhere in utils instead(and docs, please)?

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/arrays/PrimitiveNDArray.kt [hunk 4]
```diff
@@ -485,60 +485,69 @@ open class PrimitiveNDArray(array: PrimitiveTiledArray, strides: Strides) : Numb
...
         val n = actualThis.shape[0]
-//        val t = actualThis.shape[1]
+        val t = actualThis.shape[1]
+        val m = actualOther.shape[1]
+
 
         val lBlockSize = actualThis.array.blockSize
-        val rdBlockSize = destination.array.blockSize
+        val rowFlops = t * m
+        val batchSize = let {
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/arrays/PrimitiveNDArray.kt:494]: Separate function

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveExtensions.kt [hunk 5]
```diff
@@ -120,53 +120,125 @@ suspend fun PrimitiveNDArray.dequantize(zeroPoint: PrimitiveNDArray?, scale: Flo
...
-    val lrBlockSize = array.blockSize
+    val lrBlockSize = this.array.blockSize
+
+    val destBlocks = destination.array.blocks
+    val leftBlocks = this.array.blocks
+    val rightBlocks = other.array.blocks
+    val rowFlops = t * m
+    val zero = (0).toPrimitive()
+
+    val batchSize = run {
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveExtensions.kt:140]: Separate function

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt [hunk 6]
```diff
@@ -1,68 +1,206 @@
...
-    DataType.FLOAT -> object : FloatMap {
-        override fun apply(value: Float): Float = kotlin.math.exp(value)
+suspend fun softmaxFloat(input: FloatNDArray, rows: Int, columns: Int): MutableFloatNDArray {
+    val inputBlockSize = input.array.blockSize
+    val inputBlocks = input.array.blocks
+
+    val outputArray = FloatTiledArray(input.linearSize, input.array.blockSize)
+    val maxesArray = FloatArray(inputBlocks.size)
+
+    val maxBatchSize = run {
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt:17]: Separate function

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt [hunk 7]
```diff
@@ -1,68 +1,206 @@
...
-    input: NDArrayCore,
-    axis: Int = 0,
-    strides: Strides = input.strides
-): MutableNumberNDArrayCore {
-    fun resolveDims(dims: IntArray?): Int {
-        return if (dims.isNullOrEmpty()) 1 else dims!!.reduce(Int::times)
+    val blocksInRow = columns / inputBlockSize
+
+
+    val minusBatchSize = run {
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt:46]: Separate this one too

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt [hunk 8]
```diff
@@ -1,68 +1,206 @@
...
-                    for (row in i until end) {
-                        matrixRows[row].softmaxRow()
-                    }
+                    minusWrapper(rowStart, min(rowStart + minusBatchSize, rows))
                 }
             }
         }
+    }
+
+    val expBatchSize = run {
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/Softmax.kt:85]: And this

**Review by AnastasiaTuchina — CHANGES_REQUESTED** [review]

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/biasGelu/BiasGelu.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package io.kinference.ndarray.extensions.biasGelu
+
+import io.kinference.ndarray.arrays.*
+import io.kinference.primitives.types.DataType
+
+suspend fun biasGelu(input: NumberNDArrayCore, bias: NumberNDArrayCore): MutableNumberNDArrayCore {
+    require(
+        (input.type == DataType.FLOAT && bias.type == DataType.FLOAT) ||
+            (input.type == DataType.DOUBLE && bias.type == DataType.DOUBLE)
+    ) { "BiasGelu operation supported only for DoubleNDArray and FloatNDArray" }
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/biasGelu/BiasGelu.kt:10]: Maybe put the actual data types of input and bias in this error message?

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/softmax/Softmax.kt [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+import io.kinference.ndarray.arrays.*
+import io.kinference.primitives.types.DataType
+
+suspend fun softmax(
+    input: NumberNDArrayCore,
+    axis: Int = 0,
+    strides: Strides = input.strides
+): MutableNumberNDArrayCore {
+    require(input.type == DataType.FLOAT || input.type == DataType.DOUBLE) { "Softmax operation supported only for DoubleNDArray and FloatNDArray" }
```
Comment (AnastasiaTuchina) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/softmax/Softmax.kt:11]: Actual data type as well

**Review by AnastasiaTuchina — CHANGES_REQUESTED** [review]

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/Utils.kt [hunk 11]
```diff
@@ -106,3 +109,32 @@ internal fun IntArray.swap(leftIdx: Int, rightIdx: Int) {
     this[leftIdx] = this[rightIdx]
     this[rightIdx] = temp
 }
+/*
+ * Parallelize with batching by minDataPerLaunch
+ */
+suspend fun parallelizeByBlocks(blockSize: Int,
+                                countBlocks: Int,
+                                minDataPerLaunch: Int,
+                                function: (blockStart: Int, blockEnd: Int) -> Unit) {
```
Comment (TanVD) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/Utils.kt:118]: By convention, we call it "body"

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/Utils.kt [hunk 12]
```diff
@@ -106,3 +109,32 @@ internal fun IntArray.swap(leftIdx: Int, rightIdx: Int) {
...
+            for (blockStart in 0 until countBlocks step batchSize) {
+                launch {
+                    function(blockStart, min(blockStart + batchSize, countBlocks))
+                }
+            }
+        }
+    }
+}
+
+suspend inline fun parallelizeByRows(rowSize: Int, countRows: Int, minDataPerLaunch: Int, noinline function: (rowStart: Int, rowEnd: Int) -> Unit) = parallelizeByBlocks(rowSize, countRows, minDataPerLaunch, function)
```
Comment (TanVD) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/Utils.kt:140]: Why do you need suspend modifier for inline function? You can just inline internal callback as well, and then inline function will work in every code: both suspendable and non-suspendable.

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/arrays/PrimitiveNDArray.kt [hunk 13]
```diff
@@ -485,60 +483,50 @@ open class PrimitiveNDArray(array: PrimitiveTiledArray, strides: Strides) : Numb
...
+        val rdBlocksInRow = actualOther.blocksInRow
+
+        val leftBlocks = actualThis.array.blocks
+        val rightBlocks = actualOther.array.blocks
+        val destBlocks = destination.array.blocks
+        val lBlockSize = actualThis.array.blockSize
+
+        val nRowFlop = t * m
+
+        parallelizeByRows(nRowFlop, n, 261120) { nStart, nEnd ->
```
Comment (TanVD) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/arrays/PrimitiveNDArray.kt:500]: A bit of information regarding that magic constant?

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveExtensions.kt [hunk 14]
```diff
@@ -120,53 +120,54 @@ suspend fun PrimitiveNDArray.dequantize(zeroPoint: PrimitiveNDArray?, scale: Flo
...
+    val dBlockSize = destination.array.blockSize
+    val lrBlockSize = this.array.blockSize
+
+    val destBlocks = destination.array.blocks
+    val leftBlocks = this.array.blocks
+    val rightBlocks = other.array.blocks
+    val rowFlop = t * m
+    val zero = (0).toPrimitive()
+
+    parallelizeByRows(rowFlop, n, 262144) { nStart: Int, nEnd: Int ->
```
Comment (TanVD) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/PrimitiveExtensions.kt:140]: And here

File: ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/softmax/SoftmaxPrimitive.kt [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    parallelizeByBlocks(inputBlockSize, inputBlocks.size, 65536) { blockStart, blockEnd ->
+        for (blockNum in blockStart until blockEnd) {
+            maxesArray[blockNum] = inputBlocks[blockNum].max()
+        }
+    }
+
+    val blocksInRow = columns / inputBlockSize
+
+    //Minus maximum from input and store in output
+    parallelizeByRows(columns, rows, 1048576) { rowStart, rowEnd ->
```
Comment (TanVD) [inline ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/extensions/softmax/SoftmaxPrimitive.kt:33]: And here

**Review by TanVD — CHANGES_REQUESTED** [review]

**Review by TanVD — APPROVED** [review]

**Review by AnastasiaTuchina — APPROVED** [review]

**Pull request merged**